File: src/0100_same_tree.js
function isSameTree(p, q) {
    if (!p && !q) return true;
    if (p && q && p.val === q.val) return isSameTree(p.left, q.left) && isSameTree(p.right, q.right);
    return false;
}

module.exports = { isSameTree };

if (require.main === module) {
    const assert = require('assert');
    function ListNode(val = 0, next = null) { this.val = val; this.next = next; }
    function TreeNode(val = 0, left = null, right = null) { this.val = val; this.left = left; this.right = right; }
    function Node(val = 0, neighbors = []) { this.val = val; this.neighbors = neighbors; }
    function listNodeToString(node) {
        const values = [];
        let current = node;
        while (current !== null) {
            values.push(String(current.val));
            current = current.next;
        }
        return values.join('->');
    }
    function treeToString(node) {
        if (node === null) return 'None';
        return `${node.val}(${treeToString(node.left)},${treeToString(node.right)})`;
    }
    function sorted(value) { return value.slice().sort(); }
    assert.deepStrictEqual(isSameTree(new TreeNode(1, new TreeNode(2, null, null), new TreeNode(3, null, null)), new TreeNode(1, new TreeNode(2, null, null), new TreeNode(3, null, null))), true);
    assert.deepStrictEqual(isSameTree(new TreeNode(1, new TreeNode(2, null, null), null), new TreeNode(1, null, new TreeNode(2, null, null))), false);
    assert.deepStrictEqual(isSameTree(new TreeNode(1, new TreeNode(2, null, null), new TreeNode(1, null, null)), new TreeNode(1, new TreeNode(1, null, null), new TreeNode(2, null, null))), false);
}

File: src/0100_same_tree.rs
use std::cell::RefCell;
use std::rc::Rc;
struct Solution;

#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: Option<Rc<RefCell<TreeNode>>>,
    pub right: Option<Rc<RefCell<TreeNode>>>,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        Self {
            val,
            left: None,
            right: None,
        }
    }
}

impl Solution {
    pub fn is_same_tree(
        p: Option<Rc<RefCell<TreeNode>>>,
        q: Option<Rc<RefCell<TreeNode>>>,
    ) -> bool {
        match (p, q) {
            (None, None) => true,
            (Some(p), Some(q)) => {
                let p = p.borrow();
                let q = q.borrow();
                p.val == q.val
                    && Self::is_same_tree(p.left.clone(), q.left.clone())
                    && Self::is_same_tree(p.right.clone(), q.right.clone())
            }
            _ => false,
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;
    use std::cell::RefCell;
    use std::rc::Rc;

    fn node(
        val: i32,
        left: Option<Rc<RefCell<TreeNode>>>,
        right: Option<Rc<RefCell<TreeNode>>>,
    ) -> Option<Rc<RefCell<TreeNode>>> {
        Some(Rc::new(RefCell::new(TreeNode { val, left, right })))
    }

    fn tree_to_string(root: Option<Rc<RefCell<TreeNode>>>) -> String {
        match root {
            Some(node) => {
                let node = node.borrow();
                format!(
                    "{}({},{})",
                    node.val,
                    tree_to_string(node.left.clone()),
                    tree_to_string(node.right.clone())
                )
            }
            None => "None".to_string(),
        }
    }

    #[test]
    fn main_asserts() {
        assert_eq!(
            Solution::is_same_tree(
                node(1, node(2, None, None), node(3, None, None)),
                node(1, node(2, None, None), node(3, None, None))
            ),
            true
        );
        assert_eq!(
            Solution::is_same_tree(
                node(1, node(2, None, None), None),
                node(1, None, node(2, None, None))
            ),
            false
        );
        assert_eq!(
            Solution::is_same_tree(
                node(1, node(2, None, None), node(1, None, None)),
                node(1, node(1, None, None), node(2, None, None))
            ),
            false
        );
    }
}
